### ISSUE TITLE
Workflows: "Nudge" Emails (Remind if inaction) -- Due January 2021

### ISSUE THREAD

**Issue by timcortesi** [EscherLabs/Graphene#94]
Build some kind of mechanism on a workflow or a workflow state to allow nudge emails or some sort of reminders if no actions are taken within a specified period of time. 

This has been requested for multiple workflows, but the most immediate need is from Student Records, and will be required for a workflow which is slated to go live some time in January.

**Comment by timcortesi** [comment]
Front-end UI has been implemented.  Still requires back-end implementation of CRON script

**Comment by timcortesi** [comment]
Add some code to Graphene to wake up once per day at a given time (2am?)
Get a list of all Workflow Submissions which are currently Open
From there, look up the Instance Associated w/ each workflow submission.
From there, look up the version / code associated w/ each instance:
```php
$myWorkflowInstance->findVersion();
for each($myWorkflowInstance->workflow->code->flow[3]->actions) {
  // check to see if assignment->type === 'internal'  && isset(delay)
  // check to see if there has been no inactivity in X days
  // do something -- but not really, because the code you need to call is in a controller
}
```
Put all of that code in here: 
```php
WorkflowSubmissionActionController.php
//and I think this goes in the Kernel Somewhere:
$schedule->call('App\Http\Controllers\WorkflowSubmissionActionController@inactivity_cron')->everyDay();
```

**Comment by timcortesi** [comment]
Changed issue description to no longer reference "nudge emails", as the recommended implementation allows for any action based on a period of inactivity.

**Comment by alikemaltanriverdi** [comment]
Made another commit (https://github.com/EscherLabs/Graphene/commit/b1cae17d9aa7fb125149bad6ac6bd6685377d54c) that fixes the bug caused from going through each state.

**Comment by timcortesi** [comment]
This has been merged into binghamton-staging and is up-to-date as of 37f8280.  Will be pushing this to production my.binghamton.edu on 5/20/21.

**Comment by Cloverstone** [comment]
@timcortesi Should this be closed now?

**Comment by timcortesi** [comment]
@Cloverstone I still need to build an AWS friendly mechanism for triggering scheduled jobs via Cron or similar.  Should be able to borrow from work on bcomply, but this is not quite production finalized.

**Comment by timcortesi** [comment]
More info about this issue is available here: https://github.com/BinghamtonUniversity/InnovationTeam/issues/106

**Comment by timcortesi** [comment]
I believe I have a workable solution for running cron in AWS with this commit: https://github.com/EscherLabs/Graphene/commit/f6e4e482104326fb8f36079bc33b3edb7740e98d

HOWEVER... I am concerned that we should not be sending automated emails to people in DEV environments.  I think we need to do something like this: https://github.com/BinghamtonUniversity/BComply/blob/master/config/mail.php (MAIL_LIMIT_SEND)

I'm going to try to implement something like that so we don't start spamming people in DEV.

**Comment by timcortesi** [comment]
Implemented `MAIL_LIMIT_SEND` here: https://github.com/EscherLabs/Graphene/commit/a58a7c5568dfedffc5a9ff7655a4e0fd7c120b45

This should be all set now... just waiting on pushing these changes to production (after merging with remove-berry branch)

**Comment by alikemaltanriverdi** [comment]
@timcortesi is there any update on this? Please let me know if you need any additional documentation or help regarding the issue.

**Comment by timcortesi** [comment]
@alikemaltanriverdi I'm pretty sure that this is live.  The cron job is installed and is running every minute: https://github.com/EscherLabs/Graphene/blob/Merge-prod-into-removeberry-with-staging/.ebextensions/04installcron.config

Do you have reason to believe that this isn't working?

**Comment by alikemaltanriverdi** [comment]
@timcortesi I was going through all of the Graphene issues to make sure that we were not missing anything, and I wasn't sure that cron was installed, so I just want to check with you. I'll test this in the workflows to make sure that it's working as expected in prod as well. If we have a successful test in prod, I can close this issue. Thank you.

**Comment by timcortesi** [comment]
MAIL_LIMIT_SEND is set to true in dev, so I’m pretty sure you won’t get any emails from the DEV system.

You’d have to test it in production. 

-Tim

**Comment by alikemaltanriverdi** [comment]
I just set to days of inactivity to 0 in a workflow in production, but it didn't run ( there is not action log in the submission so far, not even any error log or anything), so I set it back to 1. We'll probably see the result tomorrow by this time.

**Comment by timcortesi** [comment]
https://github.com/EscherLabs/Graphene/blob/Merge-prod-into-removeberry-with-staging/config/app.php

`    'workflow_inactivity_time'=>env('WORKFLOW_INACTIVITY_TIME','2:30'),
`

It doesn't run at random times during the day.  Please look at the code to confirm how best to test this feature, and when we should/should not expect an email to be sent.

**Comment by alikemaltanriverdi** [comment]
@timcortesi That is correct, it will run at 2:30am, if there is no other WORKFLOW_INACTIVITY_TIME environment value set to another time. 
I believe that the format is 24h time format, so we'd have to add a new env variable to test it in the production, or I can let you know about the result early in the morning once the task is run at night.

**Comment by timcortesi** [comment]
For this to work correctly and run on one server (as intended, instead off across all load balanced servers), we need to add a cache driver to Graphene.  This can be done w/ mysql directly, or with redis (which is what we use for IAMBing).  I am removing the "awaiting test" flag from this issue, as this still needs to be done.  

Based on prior conversations, we had talked about using the "database" cache driver which just uses mysql, but does require an additional table to be created in the database:

```php
Schema::create('cache', function ($table) {
    $table->string('key')->unique();
    $table->text('value');
    $table->integer('expiration');
});
```

For more information, please reference:
https://laravel.com/docs/9.x/scheduling
https://laravel.com/docs/9.x/cache#prerequisites-database
